android/src/main/java/com/tailscale/ipn/QuickToggleService.java

```diff
@@ -91,7 +85,22 @@ private void onTileClick() {
         if (needsToStop) {
             app.stopVPN();
         } else {
+            Intent intent = VpnService.prepare(this);
+            if (intent != null) {
+                // VPN permission is not granted, start MainActivity to request it
+                startMainActivity();
+            }
             app.startVPN();
         }
     }
+
+    private void startMainActivity() {
+        Intent i = getPackageManager().getLaunchIntentForPackage(getPackageName());
```

Comment:
Is there a broader issue here in that some of the VPN prep is dependent on the main activity getting started?   Can we decouple those things so that the burden is not on all intents or other activities to launch the main activity?

Reply:
Thanks for pointing this out. Moved VPN prep into App